### PR TITLE
Ignore StaleSymbol exceptions when testing whether dependencies shoul…

### LINKED ISSUE #13532: StaleSymbol error caused by macro and import
## Compiler version

3.0.2

## Minimized code

TestMacro.scala
```Scala
package testcode

import scala.quoted.Quotes

object TestMacro {
    private def impl()(using Quotes) = '{ 123 }
    inline def call(): Int = ${ impl() }
}
```

Foo.scala
```Scala
package testcode

class Foo {
	TestMacro.call()
}
```

Bar.scala
```Scala
package testcode

import testcode.Foo

class Bar(f: Foo) {
	TestMacro.call()
}

```

## Output (click arrow to expand)
<details>

```scala
[error] ## Exception when compiling 3 sources to E:\Programmieren\scala3_sbt_test\target\scala-3.0.2\classes
[error] dotty.tools.dotc.core.Denotations$StaleSymbol: stale symbol; module class Foo$#2732 in module class testcode, defined in Period(1..10, run = 2), is referred to in run Period(3..3, run = 3)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.staleSymbolError(Denotations.scala:955)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.bringForward(Denotations.scala:754)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.toNewRun$1(Denotations.scala:803)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:122)
[error] dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:116)
[error] dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:109)
[error] dotty.tools.dotc.core.Symbols$ClassSymbol.classDenot(Symbols.scala:457)
[error] dotty.tools.dotc.core.Symbols$.toClassDenot(Symbols.scala:497)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.isAbsent(SymDenotations.scala:568)
[error] dotty.tools.dotc.sbt.ExtractDependenciesCollector.dotty$tools$dotc$sbt$ExtractDependenciesCollector$$ignoreDependency(ExtractDependencies.scala:344)
[error] dotty.tools.dotc.sbt.ExtractDependenciesCollector.dotty$tools$dotc$sbt$ExtractDependenciesCollector$$addMemberRefDependency(ExtractDependencies.scala:314)
[error] dotty.tools.dotc.sbt.ExtractDependenciesCollector.addImported$1(ExtractDependencies.scala:362)
[error] dotty.tools.dotc.sbt.ExtractDependenciesCollector.traverse$$anonfun$2(ExtractDependencies.scala:366)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:889)
[error] dotty.tools.dotc.sbt.ExtractDependenciesCollector.traverse(ExtractDependencies.scala:368)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1637)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1637)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.fold$1(Trees.scala:1512)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply(Trees.scala:1514)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1613)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1638)
[error] dotty.tools.dotc.sbt.ExtractDependenciesCollector.traverse(ExtractDependencies.scala:406)
[error] dotty.tools.dotc.sbt.ExtractDependencies.run(ExtractDependencies.scala:69)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:231)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:172)
[error] dotty.tools.dotc.Driver.finish(Driver.scala:60)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:40)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.util.concurrent.FutureTask.run(Unknown Source)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
[error] java.util.concurrent.FutureTask.run(Unknown Source)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[error] java.lang.Thread.run(Unknown Source)
[error]
[error] dotty.tools.dotc.core.Denotations$StaleSymbol: stale symbol; module class Foo$#2732 in module class testcode, defined in Period(1..10, run = 2), is referred to in run Period(3..3, run = 3)
[error]         at dotty.tools.dotc.core.Denotations$SingleDenotation.staleSymbolError(Denotations.scala:955)
[error]         at dotty.tools.dotc.core.Denotations$SingleDenotation.bringForward(Denotations.scala:754)
[error]         at dotty.tools.dotc.core.Denotations$SingleDenotation.toNewRun$1(Denotations.scala:803)
[error]         at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error]         at dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:122)
[error]         at dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:116)
[error]         at dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:109)
[error]         at dotty.tools.dotc.core.Symbols$ClassSymbol.classDenot(Symbols.scala:457)
[error]         at dotty.tools.dotc.core.Symbols$.toClassDenot(Symbols.scala:497)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.isAbsent(SymDenotations.scala:568)
[error]         at dotty.tools.dotc.sbt.ExtractDependenciesCollector.dotty$tools$dotc$sbt$ExtractDependenciesCollector$$ignoreDependency(ExtractDependencies.scala:344)
[error]         at dotty.tools.dotc.sbt.ExtractDependenciesCollector.dotty$tools$dotc$sbt$ExtractDependenciesCollector$$addMemberRefDependency(ExtractDependencies.scala:314)
[error]         at dotty.tools.dotc.sbt.ExtractDependenciesCollector.addImported$1(ExtractDependencies.scala:362)
[error]         at dotty.tools.dotc.sbt.ExtractDependenciesCollector.traverse$$anonfun$2(ExtractDependencies.scala:366)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error]         at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error]         at scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error]         at scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:889)
[error]         at dotty.tools.dotc.sbt.ExtractDependenciesCollector.traverse(ExtractDependencies.scala:368)
[error]         at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1637)
[error]         at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1637)
[error]         at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.fold$1(Trees.scala:1512)
[error]         at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply(Trees.scala:1514)
[error]         at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1613)
[error]         at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1638)
[error]         at dotty.tools.dotc.sbt.ExtractDependenciesCollector.traverse(ExtractDependencies.scala:406)
[error]         at dotty.tools.dotc.sbt.ExtractDependencies.run(ExtractDependencies.scala:69)
[error]         at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error]         at scala.collection.immutable.List.map(List.scala:246)
[error]         at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error]         at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error]         at dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
[error]         at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error]         at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error]         at dotty.tools.dotc.Run.compileUnits(Run.scala:231)
[error]         at dotty.tools.dotc.Run.compileUnits(Run.scala:172)
[error]         at dotty.tools.dotc.Driver.finish(Driver.scala:60)
[error]         at dotty.tools.dotc.Driver.doCompile(Driver.scala:40)
[error]         at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error]         at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error]         at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error]         at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error]         at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
[error]         at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
[error]         at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error]         at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error]         at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error]         at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error]         at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error]         at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error]         at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error]         at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error]         at sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error]         at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:46)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:28)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:23)
[error]         at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error]         at sbt.Execute.work(Execute.scala:291)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error]         at java.util.concurrent.FutureTask.run(Unknown Source)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
[error]         at java.util.concurrent.FutureTask.run(Unknown Source)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[error]         at java.lang.Thread.run(Unknown Source)
[error] (Compile / compileIncremental) dotty.tools.dotc.core.Denotations$StaleSymbol: stale symbol; module class Foo$#2732 in module class testcode, defined in Period(1..10, run = 2), is referred to in run Period(3..3, run = 3)
```
</details>

### PR DESCRIPTION
**Pull request by odersky**
…d be ignored

Fixes #13532

A question is whether the catch should be in `ignoreDependency` in the SBT dependency
extractor or in `isAbsent` itself. I put it in `ignoreDependency` for now in order
not to hide possible stale symbol errors since isAbsent is called quite often.

### CODE REVIEW

**Comment by smarter** [comment]
Catching in ignoreDependency seems safer indeed, but I don't understand a good grasp on how suspended compilation works: why do we get a stale symbol at all here?

**Comment by odersky** [comment]
The stale symbol errors involve constructor proxies. I believe it's because we are generating them later on demand that the normal symbol update mechanisms don't always work. I am not sure about the precise details and don't have time to dig into it more. In the case of suspended computation it looks like it's the two phases that systematically follow all symbol links in usages that hit the stale symbol error. I verified that a program that calls a constructor proxy directly, but uses neither extract option, compiles correctly.

@bishabosha Can you check that the updated semanticdb expect files are reasonable?

File: compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala [hunk 1]
```diff
@@ -241,8 +242,13 @@ class ExtractSemanticDB extends Phase:
               if imported != nme.WILDCARD then
                 for alt <- tree.expr.tpe.member(imported).alternatives do
                   registerUseGuarded(None, alt.symbol, sel.imported.span, tree.source)
-                  if (alt.symbol.companionClass.exists)
-                    registerUseGuarded(None, alt.symbol.companionClass, sel.imported.span, tree.source)
+                  try
+                    if (alt.symbol.companionClass.isDefinedInCurrentRun)
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala:246]: Why was this changed to `.isDefinedInCurrentRun` from `.exists`? the test case `pos/i13532` still passes if i change it back, and then the semanticdb expect files do not need to change

**Review by bishabosha — APPROVED** [review]
semanticdb changes LGTM

File: compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala (hunk 1 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala]: resolved by newer commits

**Pull request merged**